### PR TITLE
Feature/16 furigoma

### PR DESCRIPTION
**Pull request by bradychen2**
# Description
- exect furigoma to determine which player goes first

# How to test
- API route: POST /gungi/:gungiId/furigoma
- request
```
body: { playerId: 'string' }
```
- response
```
{
        name: 'Furigoma',
        data: {
          turn, // TURN.FIRST or TURN.SECOND
          result: fiveRandomsNum, // ex. [1, 1, 1, 0, 0]
        },
}
```
# New Dependencies
N/A

### CODE REVIEW

File: server/src/gateway/controllers/Gungi.controller.ts [hunk 1]
```diff
@@ -3,24 +3,21 @@ import FurigomaUsecase from '../../usecases/service-class/FurigomaUsecase';
...
-    const request = {
-      gungiId: id,
-      body,
-    };
-
-    const response = await this._furigomaUsecase.execute();
+  @Post('/gungi/:gungiId/furigoma')
+  async furigoma(@Param('gungiId') gungiId: string, @Body() body: any) {
+    const presenter = new FurigomaPresenter();
+    const input = new FurigomaUsecaseInput(gungiId, body.playerId);
```
Comment (zhihdd) [inline server/src/gateway/controllers/Gungi.controller.ts:19]: input 可以討論一下要不要統一

File: server/src/domain/Gungi.ts [hunk 2]
```diff
@@ -112,14 +114,76 @@ class Gungi {
...
+      return 1;
+    });
+  }
+
+  async furigoma(player: Player, opponent: Player): Promise<Event[]> {
+    try {
+      // five random number range 0 - 1
+      const fiveRandomsNum = await this.genFiveRandomsNum();
+      const sum = fiveRandomsNum.reduce((acc, num) => (acc += num), 0);
+      const turn: TURN = this.determineTurn(sum);
```
Comment (zhihdd) [inline server/src/domain/Gungi.ts:171]: - await ?
- five是細節未必要表示出來
- 其實這個階層在乎的determineTurn的結果

File: server/src/domain/Gungi.ts [hunk 3]
```diff
@@ -112,14 +114,76 @@ class Gungi {
...
+      player.side = turn === TURN.FIRST ? SIDE.BLACK : SIDE.WHITE;
+      opponent.side = turn === TURN.FIRST ? SIDE.WHITE : SIDE.BLACK;
```
Comment (zhihdd) [inline server/src/domain/Gungi.ts:173]: TURN.FIRST 這是什麼意思?!

File: server/test/unitTest/furigoma.spec.ts [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+        .mockReturnValueOnce(0.3)
+        .mockReturnValueOnce(0.5)
+        .mockReturnValueOnce(0.6);
+      const result: (0 | 1)[] = await gungi['genFiveRandomsNum']();
+
+      // Then
+      expect(spyRandom).toBeCalledTimes(5);
+      expect(result).toEqual([0, 0, 0, 1, 1]);
+    });
+  });
```
Comment (zhihdd) [inline server/test/unitTest/furigoma.spec.ts:101]: private method 會需要驗嗎?

File: server/test/unitTest/furigoma.spec.ts [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+    it('躑駒正面 < 3，房主為後手玩家，回傳 TURN.SECOND', async () => {
+      // Given
+      const gungi = new Gungi('test', LEVEL.BEGINNER, [], new GungiHan());
+
+      // When 正面 < 3
+      const result = await gungi['determineTurn'](2);
+
+      // Then
+      expect(result).toBe(TURN.SECOND);
+    });
```
Comment (zhihdd) [inline server/test/unitTest/furigoma.spec.ts:124]: 同上

File: server/test/unitTest/furigoma.spec.ts [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+import Gungi from '../../src/domain/Gungi';
+import GungiHan from '../../src/domain/GungiHan';
+import TURN from '../../src/domain/constant/TURN';
+
+describe('Furigoma - 躑駒', () => {
+  describe('多數正面 - 房主先手', () => {
+    it('房主為玩家 A，躑駒正面 >= 3，房主為先手玩家', async () => {
+      // Given
+      // A 為房主
+      const playerA = new Player(
```
Comment (zhihdd) [inline server/test/unitTest/furigoma.spec.ts:15]: given / when / then  可以各自抽成方法讓他有語意

File: server/src/gateway/presenter/FurigomaPresenter.ts [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+    const targetEvents: Event[] = [];
+
+    events.forEach((event) => {
+      if (event.name === 'Furigoma') {
+        targetEvents.push(event);
+      }
+    });
+    return targetEvents;
+  }
+}
```
[the remark this answers is not in the document]
Comment (bradychen2) [reply server/src/gateway/presenter/FurigomaPresenter.ts]: 漏掉哩，重推補上

File: server/src/gateway/controllers/Gungi.controller.ts (hunk 1 above)
Comment (bradychen2) [reply server/src/gateway/controllers/Gungi.controller.ts:23]: 統一成 request naming

File: server/src/domain/Gungi.ts [hunk 8]
```diff
@@ -112,14 +114,76 @@ class Gungi {
...
+    }
+  }
+
+  private genFiveRandomsNum(): (1 | 0)[] {
+    return [
+      Math.random(),
+      Math.random(),
+      Math.random(),
+      Math.random(),
+      Math.random(),
```
[the remark this answers is not in the document]
Comment (bradychen2) [reply server/src/domain/Gungi.ts:157]: 重推改其成比較簡潔的寫法

File: server/src/domain/Gungi.ts [hunk 9]
```diff
@@ -112,14 +114,76 @@ class Gungi {
...
+    ].map((num) => {
+      if (num < 0.5) {
```
[the remark this answers is not in the document]
Comment (bradychen2) [reply server/src/domain/Gungi.ts:159]: 想不到怎麼命名，這裡要表達的就是 50-50 的機率正反面，或是用註解說明？

[the remark this answers is not in the document]
Comment (zhihdd) [reply server/src/domain/Gungi.ts:159]: 我覺得用變數percentage就可以了

File: server/src/domain/Gungi.ts (hunk 2 above)
Comment (bradychen2) [reply server/src/domain/Gungi.ts:171]: 1. 刪除 await 
2-1. 能想到保留的好處：一看函式就知道 domain 的規則是用 5 個兵去決定先後順序
2-2. 刪掉的好處：規則改變 (ex. 躑駒要改用 6, 7, 8 個兵去決定) 不用改函式名稱，但要進到內容詳讀

File: server/src/domain/Gungi.ts (hunk 3 above)
Comment (bradychen2) [reply server/src/domain/Gungi.ts:173]: 改成 TURN.SENTE TURN.GOTE

File: server/test/unitTest/furigoma.spec.ts (hunk 4 above)
Comment (bradychen2) [reply server/test/unitTest/furigoma.spec.ts:101]: private method 不需要驗證的理論是什麼？因為驗過 Acceptance 所以 private 都會是正確的？

File: server/test/unitTest/furigoma.spec.ts (hunk 5 above)
Comment (bradychen2) [reply server/test/unitTest/furigoma.spec.ts:124]: 同上

File: server/test/unitTest/furigoma.spec.ts (hunk 4 above)
Comment (zhihdd) [reply server/test/unitTest/furigoma.spec.ts:101]: private mehtod 可以說是，public method 內部的細節出來的東西，既然如此，我驗了public不就連private都一起驗了，那為何要再另外拉出來驗?
或者是說是什麼原因讓你對特定的private method不具有信心導致你想要去驗他

File: server/src/domain/Gungi.ts (hunk 2 above)
Comment (zhihdd) [reply server/src/domain/Gungi.ts:171]: fiveRandomNum 覺得在furigoma看到第一眼我無法理解到底要幹嘛，但如果包在determineTurn內，透過這層抽象的意義，可以知道fiveRandomNum 和 sum是determineTurn的實作細節，至於你說的"抽換性"，並不是我想要你這樣做的原因，主要是為了"更好的可讀性"，"抽換性"只是附帶好處

File: server/test/unitTest/furigoma.spec.ts (hunk 4 above)
Comment (bradychen2) [reply server/test/unitTest/furigoma.spec.ts]: 產生隨機數(五個正反)的函式在外層 public method 驗的時候是被 mock 掉的，所以有可能哪天有人去動了 private function，外層函式還是驗的過的情況。所以這個狀況 private func 要被驗證。

File: server/src/domain/Gungi.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (bradychen2) [reply server/src/domain/Gungi.ts:158]: 改用變數

File: server/src/domain/Gungi.ts [hunk 10]
```diff
@@ -112,14 +114,76 @@ class Gungi {
...
+        return 0;
+      }
+      return 1;
```
[the remark this answers is not in the document]
Comment (bradychen2) [reply server/src/domain/Gungi.ts]:
```
enum FURIGOMA {
  HEADS = 1,
  TAILS = 0,
}
```

File: server/test/unitTest/furigoma.spec.ts (hunk 5 above)
Comment (bradychen2) [reply server/test/unitTest/furigoma.spec.ts]: 這個函式跟上面的情況不太一樣，外層呼叫的時候這個 private method 沒有被 mock，所以的確可以不個別驗。不過更複雜的 context 下會不會有外層測試沒有涵蓋到呼叫的 private function 所有 edge cases 的情況？

File: server/src/domain/Gungi.ts (hunk 2 above)
Comment (bradychen2) [reply server/src/domain/Gungi.ts]: fiveRandomNums 改成 genTossResult，tossResult 還要回傳到外部所以我不把他包到 determineTurn 裏面。sum 當作 determineTurn 的細節合適，所以移進去。

File: server/src/domain/GungiHan.ts [hunk 11]
```diff
@@ -28,6 +36,14 @@ class GungiHan {
...
     this._han = map;
   }
+
+  updateHan(goma: Goma, to: Coordinate) {
+    const old = this._han
+      .get(goma.getCoordinateX())
+      .get(goma.getCoordinateY())
+      .pop();
+    this._han.get(to.x).get(to.y).push(goma);
+  }
```
[the remark this answers is not in the document]
Comment (bradychen2) [reply server/src/domain/GungiHan.ts]: merge with feature/16-configuration, delete this method

**Review by allyusd — CHANGES_REQUESTED** [review]
除了原本小智留的 `given / when / then 可以各自抽成方法讓他有語意` 之外，
還有兩個可以刪除的程式碼，修掉我這邊就 ok

File: server/test/unitTest/furigoma.spec.ts (hunk 6 above)
Comment (bradychen2) [reply server/test/unitTest/furigoma.spec.ts:15]: done

File: server/src/domain/Gungi.ts [hunk 12]
```diff
@@ -211,26 +256,22 @@ class Gungi {
...
-          this._goteDeadArea = player.deadArea;
-          this._goteGomaOki = player.gomaOki;
-          break;
-        }
-        default: {
-          throw new Error('沒有這個玩家');
-        }
-      }
-    });
+  // setSenteGote(players: Player[]) {
```
[the remark this answers is not in the document]
Comment (bradychen2) [reply server/src/domain/Gungi.ts]: done

File: server/test/e2e/e2e.e2e-spec.ts [hunk 13]
```diff
@@ -83,4 +96,114 @@ describe('AppController (e2e)', () => {
       .expect(200)
       .expect({ winner: 'B' });
   });
+
+  it('/(POST) gungi/:gungiId/configuration', async () => {
```
[the remark this answers is not in the document]
Comment (bradychen2) [reply server/test/e2e/e2e.e2e-spec.ts]: done

**Review by allyusd — APPROVED** [review]

File: server/src/domain/Gungi.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (bradychen2) [reply server/src/domain/Gungi.ts]: done

File: server/test/e2e/e2e.e2e-spec.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (bradychen2) [reply server/test/e2e/e2e.e2e-spec.ts]: done
